File: frontend/src/types/emotion.ts
export type EmotionType =
  | "happy"
  | "sad"
  | "angry"
  | "fear"
  | "surprise"
  | "disgust"
  | "neutral";

export type MoodCategory = "positive" | "neutral" | "negative";

export type EngagementLevel = "high" | "medium" | "low";

export type ContextMode = "education" | "healthcare";

export interface FaceBoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface ModalityResultBase {
  modality: "face" | "audio" | "text";
  emotion: EmotionType;
  confidence: number;
  probabilities: Record<EmotionType, number>;
  mood_category: MoodCategory;
  is_fallback: boolean;
  model_name: string;
  metadata?: Record<string, unknown>;
}

export interface FaceAnalysisResponse extends ModalityResultBase {
  face_detected: boolean;
  bounding_box: FaceBoundingBox | null;
  processed_image_base64: string | null;
}

export interface AudioAnalysisResponse extends ModalityResultBase {
  waveform_points: number[] | null;
  duration_seconds: number | null;
  sample_rate: number | null;
  pitch_estimate: number | null;
  energy_level: number | null;
}

export interface TextAnalysisResponse extends ModalityResultBase {
  character_count: number;
  word_count: number;
  sentiment_polarity: number;
}

export interface RecommendationItem {
  title: string;
  description: string;
  category: string;
  action_type: "activity" | "mindfulness" | "resource" | "alert";
}

export interface MultimodalFusionRequest {
  face_result?: FaceAnalysisResponse | null;
  audio_result?: AudioAnalysisResponse | null;
  text_result?: TextAnalysisResponse | null;
  context_mode: ContextMode;
  custom_weights?: Record<string, number>;
}

export interface MultimodalFusionResponse {
  final_emotion: EmotionType;
  confidence: number;
  engagement_level: EngagementLevel;
  mood_category: MoodCategory;
  modalities_used: ("face" | "audio" | "text")[];
  effective_weights: Record<string, number>;
  fusion_probabilities: Record<EmotionType, number>;
  agreement_score: number;
  is_mixed_emotion: boolean;
  disagreement_detected: boolean;
  context_mode: ContextMode;
  recommendations: RecommendationItem[];
  medical_disclaimer: string;
}

export interface AnalysisRecord {
  id: number;
  session_id: string;
  timestamp: string;
  context_mode: ContextMode;
  face_emotion?: EmotionType;
  face_confidence?: number;
  audio_emotion?: EmotionType;
  audio_confidence?: number;
  text_emotion?: EmotionType;
  text_confidence?: number;
  final_emotion: EmotionType;
  confidence: number;
  engagement_level: EngagementLevel;
  mood_category: MoodCategory;
  agreement_score: number;
  modalities_used: ("face" | "audio" | "text")[];
  notes?: string;
}

export interface AnalysisHistoryStats {
  total_analyses: number;
  most_frequent_emotion: string;
  average_confidence: number;
  emotion_distribution: Record<string, number>;
  mood_distribution: Record<string, number>;
  engagement_distribution: Record<string, number>;
  modalities_frequency: Record<string, number>;
}

export const EMOTION_META: Record<
  EmotionType,
  { label: string; emoji: string; color: string; bgBadge: string }
> = {
  happy: {
    label: "Happy",
    emoji: "😊",
    color: "#10b981",
    bgBadge: "bg-emerald-500/20 text-emerald-300 border-emerald-500/30",
  },
  sad: {
    label: "Sad",
    emoji: "😢",
    color: "#60a5fa",
    bgBadge: "bg-blue-500/20 text-blue-300 border-blue-500/30",
  },
  angry: {
    label: "Angry",
    emoji: "😠",
    color: "#ef4444",
    bgBadge: "bg-rose-500/20 text-rose-300 border-rose-500/30",
  },
  fear: {
    label: "Fear",
    emoji: "😨",
    color: "#a855f7",
    bgBadge: "bg-purple-500/20 text-purple-300 border-purple-500/30",
  },
  surprise: {
    label: "Surprise",
    emoji: "😲",
    color: "#f59e0b",
    bgBadge: "bg-amber-500/20 text-amber-300 border-amber-500/30",
  },
  disgust: {
    label: "Disgust",
    emoji: "🤢",
    color: "#84cc16",
    bgBadge: "bg-lime-500/20 text-lime-300 border-lime-500/30",
  },
  neutral: {
    label: "Neutral",
    emoji: "😐",
    color: "#94a3b8",
    bgBadge: "bg-slate-500/20 text-slate-300 border-slate-500/30",
  },
};

File: backend/app/schemas/emotion.py
from enum import Enum
from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class EmotionType(str, Enum):
    HAPPY = "happy"
    SAD = "sad"
    ANGRY = "angry"
    FEAR = "fear"
    SURPRISE = "surprise"
    DISGUST = "disgust"
    NEUTRAL = "neutral"


class MoodCategory(str, Enum):
    POSITIVE = "positive"
    NEUTRAL = "neutral"
    NEGATIVE = "negative"


class EngagementLevel(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class ContextMode(str, Enum):
    EDUCATION = "education"
    HEALTHCARE = "healthcare"


class FaceAnalysisRequest(BaseModel):
    image_base64: Optional[str] = Field(None, description="Base64 encoded image data (data:image/...;base64,...)")


class FaceBoundingBox(BaseModel):
    x: int
    y: int
    width: int
    height: int


class ModalityResult(BaseModel):
    modality: str  # "face", "audio", "text"
    emotion: str
    confidence: float
    probabilities: Dict[str, float]
    mood_category: MoodCategory
    is_fallback: bool = False
    model_name: str = "Standard Model"
    metadata: Optional[Dict[str, object]] = None


class FaceAnalysisResponse(ModalityResult):
    face_detected: bool = False
    bounding_box: Optional[FaceBoundingBox] = None
    processed_image_base64: Optional[str] = None


class AudioAnalysisResponse(ModalityResult):
    waveform_points: Optional[List[float]] = None
    duration_seconds: Optional[float] = None
    sample_rate: Optional[int] = None
    pitch_estimate: Optional[float] = None
    energy_level: Optional[float] = None


class TextAnalysisRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=5000, description="Text to analyze for emotional sentiment")


class TextAnalysisResponse(ModalityResult):
    character_count: int
    word_count: int
    sentiment_polarity: float = 0.0


class MultimodalFusionRequest(BaseModel):
    face_result: Optional[FaceAnalysisResponse] = None
    audio_result: Optional[AudioAnalysisResponse] = None
    text_result: Optional[TextAnalysisResponse] = None
    context_mode: ContextMode = ContextMode.EDUCATION
    custom_weights: Optional[Dict[str, float]] = None


class RecommendationItem(BaseModel):
    title: str
    description: str
    category: str
    action_type: str  # e.g., "activity", "mindfulness", "resource", "alert"


class MultimodalFusionResponse(BaseModel):
    final_emotion: str
    confidence: float
    engagement_level: EngagementLevel
    mood_category: MoodCategory
    modalities_used: List[str]
    effective_weights: Dict[str, float]
    fusion_probabilities: Dict[str, float]
    agreement_score: float
    is_mixed_emotion: bool = False
    disagreement_detected: bool = False
    context_mode: ContextMode
    recommendations: List[RecommendationItem]
    medical_disclaimer: str = (
        "EmotionVerse AI provides supportive insights only and is not a medical diagnostic system."
    )
